### PR TITLE
Refactor policies to avoid unintended usage bugs

### PR DESCRIPTION
**Pull request by alexandrebodin**
### What does it do?

Refactor policies to have a consistent factory API like middlewares

### Why is it needed?

This fixes an issue with policy optionally having configs that could create bugs if misused. Now all policies are created through a factory and have the same factory signature as middlewares for ease of use and consistency.

ping @pwizla for the doc :)

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
Is there a difference in policy ctx over just normal ctx?

**Comment by pwizla** [comment]
> Is there a difference in policy ctx over just normal ctx?

From what I understand, policyCtx extends Koa's ctx (see [Alex's comment on my PR](https://github.com/strapi/documentation/pull/450#discussion_r742178481))

**Review by Elhebert — DISMISSED** [review]

File: examples/getstarted/src/policies/test-policy.js [hunk 1]
```diff
@@ -1,3 +1,18 @@
-module.exports = ctx => {
-  return true;
+'use strict';
+
+/**
+ * `coucou-policy` policy.
```
Comment (Elhebert) [inline examples/getstarted/src/policies/test-policy.js:4]: Maybe replace `coucou` by `test` to match the filename?

**Review by Elhebert — CHANGES_REQUESTED** [review]
LGTM, just a minor comment update for consistency.

File: examples/getstarted/src/policies/test-policy.js (hunk 1 above)
Comment (alexandrebodin) [reply examples/getstarted/src/policies/test-policy.js:4]: 👍

**Review by Elhebert — APPROVED** [review]
👌

**Comment by alexandrebodin** [comment]
@pwizla  this refacto has been merged and will impact the policies doc. Is the code enough to update the doc or do you want to discuss it ?

**Comment by pwizla** [comment]
> @pwizla this refacto has been merged and will impact the policies doc. Is the code enough to update the doc or do you want to discuss it ?

Thanks for the heads-up, @alexandrebodin ! I think it's enough, I've just created [this docs PR](https://github.com/strapi/documentation/pull/477), please let me know what you think :-)

**Pull request merged**